File: backend/app/external/kakao_api.py
"""
카카오 REST API 클라이언트
- Local API: 장소 검색 (은행 등)
- Navi API: 길찾기 (경로 좌표)
"""
import os
import httpx

KAKAO_REST_KEY = os.getenv("KAKAO_REST_API_KEY", "04a02b133b0d9ce03375340091c37c97")


def _headers() -> dict:
    return {"Authorization": f"KakaoAK {KAKAO_REST_KEY}"}


async def search_nearby_places(
    query: str,
    lng: float,
    lat: float,
    radius: int = 2000,
    category_group_code: str = "",
    size: int = 10,
) -> list[dict]:
    """카카오 로컬 API - 키워드 장소 검색"""
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            params = {
                "query": query,
                "x": str(lng),
                "y": str(lat),
                "radius": str(radius),
                "size": str(size),
                "sort": "distance",
            }
            if category_group_code:
                params["category_group_code"] = category_group_code

            resp = await client.get(
                "https://dapi.kakao.com/v2/local/search/keyword.json",
                params=params,
                headers=_headers(),
            )
            if resp.status_code == 200:
                data = resp.json()
                return [
                    {
                        "id": doc.get("id", ""),
                        "name": doc.get("place_name", ""),
                        "address": doc.get("address_name", ""),
                        "road_address": doc.get("road_address_name", ""),
                        "lat": float(doc.get("y", "0")),
                        "lng": float(doc.get("x", "0")),
                        "phone": doc.get("phone", ""),
                        "distance": int(doc.get("distance", "0")),
                        "category": doc.get("category_name", ""),
                    }
                    for doc in data.get("documents", [])
                ]
    except Exception:
        pass
    return []


async def search_nearby_banks(lng: float, lat: float, radius: int = 3000) -> list[dict]:
    """근처 은행 검색 (카테고리: BK9, ATM 제외)"""
    results = await search_nearby_places("은행", lng, lat, radius, category_group_code="BK9", size=15)
    # ATM, 365코너 등 무인 시설 제외 — 실제 영업점만
    exclude_keywords = ["ATM", "365", "자동화", "무인"]
    filtered = [b for b in results if not any(kw in b["name"] for kw in exclude_keywords)]
    return filtered if filtered else results[:5]


async def search_nearby_post_offices(lng: float, lat: float, radius: int = 3000) -> list[dict]:
    """근처 우체국 검색"""
    return await search_nearby_places("우체국", lng, lat, radius, size=10)


async def get_directions(
    origin_lng: float, origin_lat: float,
    dest_lng: float, dest_lat: float,
    waypoints: list[tuple[float, float]] | None = None,
) -> dict | None:
    """
    카카오 내비 API - 경로 탐색
    반환: distance(m), duration(s), route_coords(경로 좌표 리스트)
    """
    try:
        async with httpx.AsyncClient(timeout=15) as client:
            params = {
                "origin": f"{origin_lng},{origin_lat}",
                "destination": f"{dest_lng},{dest_lat}",
                "priority": "RECOMMEND",
            }
            if waypoints:
                wp_str = "|".join(f"{lng},{lat}" for lng, lat in waypoints)
                params["waypoints"] = wp_str

            resp = await client.get(
                "https://apis-navi.kakaomobility.com/v1/directions",
                params=params,
                headers=_headers(),
            )
            if resp.status_code == 200:
                data = resp.json()
                routes = data.get("routes", [])
                if not routes:
                    return None

                route = routes[0]
                summary = route.get("summary", {})

                # 경로 좌표 추출
                coords = []
                for section in route.get("sections", []):
                    for road in section.get("roads", []):
                        vertexes = road.get("vertexes", [])
                        for i in range(0, len(vertexes), 2):
                            if i + 1 < len(vertexes):
                                coords.append({
                                    "lng": vertexes[i],
                                    "lat": vertexes[i + 1],
                                })

                return {
                    "distance": summary.get("distance", 0),  # meters
                    "duration": summary.get("duration", 0),   # seconds
                    "route_coords": coords,
                }
    except Exception:
        pass
    return None


async def geocode_address(address: str) -> dict | None:
    """
    주소 → 좌표 변환 (카카오 Local 주소검색 API)
    Returns: {address, road_address, lat, lng} or None
    """
    if not address:
        return None
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(
                "https://dapi.kakao.com/v2/local/search/address.json",
                params={"query": address},
                headers=_headers(),
            )
            if resp.status_code != 200:
                return None
            docs = resp.json().get("documents", [])
            if not docs:
                # 주소 검색 실패 시 키워드 검색으로 폴백
                resp2 = await client.get(
                    "https://dapi.kakao.com/v2/local/search/keyword.json",
                    params={"query": address, "size": "1"},
                    headers=_headers(),
                )
                if resp2.status_code != 200:
                    return None
                docs2 = resp2.json().get("documents", [])
                if not docs2:
                    return None
                d = docs2[0]
                return {
                    "address": d.get("address_name", address),
                    "road_address": d.get("road_address_name", ""),
                    "lat": float(d.get("y", "0")),
                    "lng": float(d.get("x", "0")),
                }
            doc = docs[0]
            road = doc.get("road_address") or {}
            return {
                "address": doc.get("address_name", address),
                "road_address": road.get("address_name", ""),
                "lat": float(doc.get("y", "0")),
                "lng": float(doc.get("x", "0")),
            }
    except Exception:
        return None


async def reverse_geocode(lat: float, lng: float) -> dict | None:
    """
    좌표 → 주소 변환 (카카오 Local 좌표→주소 API)
    Returns: {address, road_address, region_1depth, region_2depth, region_3depth}
    """
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(
                "https://dapi.kakao.com/v2/local/geo/coord2address.json",
                params={"x": str(lng), "y": str(lat)},
                headers=_headers(),
            )
            if resp.status_code != 200:
                return None
            docs = resp.json().get("documents", [])
            if not docs:
                return None
            d = docs[0]
            addr = d.get("address") or {}
            road = d.get("road_address") or {}
            return {
                "address": addr.get("address_name", ""),
                "road_address": road.get("address_name", ""),
                "region_1depth": addr.get("region_1depth_name", ""),
                "region_2depth": addr.get("region_2depth_name", ""),
                "region_3depth": addr.get("region_3depth_name", ""),
            }
    except Exception:
        return None


async def get_multi_stop_route(
    start: tuple[float, float],
    stops: list[tuple[float, float]],
) -> list[dict]:
    """
    여러 경유지 경로를 구간별로 조회
    start: (lng, lat), stops: [(lng, lat), ...]
    반환: 구간별 [{distance, duration, route_coords}, ...]
    """
    segments = []
    current = start
    for stop in stops:
        result = await get_directions(current[0], current[1], stop[0], stop[1])
        if result:
            segments.append(result)
        else:
            segments.append({
                "distance": 0,
                "duration": 0,
                "route_coords": [
                    {"lng": current[0], "lat": current[1]},
                    {"lng": stop[0], "lat": stop[1]},
                ],
            })
        current = stop
    return segments

File: frontend/src/external/kakaoApi.ts
/**
 * Kakao REST API 클라이언트 (TS 포팅)
 *   - Local API: 장소/주소 검색
 *   - Navi API: 길찾기
 */
import { httpGet } from '../lib/httpClient'
import { API_KEYS } from '../config/apiKeys'

const LOCAL_BASE = 'https://dapi.kakao.com/v2/local'
const NAVI_BASE = 'https://apis-navi.kakaomobility.com/v1/directions'

function authHeaders(): Record<string, string> {
  return { Authorization: `KakaoAK ${API_KEYS.KAKAO_REST}` }
}

export interface KakaoPlace {
  id: string
  name: string
  address: string
  road_address: string
  lat: number
  lng: number
  phone: string
  distance: number
  category: string
}

export async function searchNearbyPlaces(params: {
  query: string
  lat: number
  lng: number
  radius?: number
  category_group_code?: string
  size?: number
}): Promise<KakaoPlace[]> {
  const { query, lat, lng, radius = 2000, category_group_code, size = 10 } = params
  try {
    const res = await httpGet<any>({
      url: `${LOCAL_BASE}/search/keyword.json`,
      params: {
        query,
        x: String(lng),
        y: String(lat),
        radius: String(radius),
        size: String(size),
        sort: 'distance',
        ...(category_group_code ? { category_group_code } : {}),
      },
      headers: authHeaders(),
    })
    if (res.status !== 200) return []
    const docs = res.data?.documents || []
    return docs.map((d: any) => ({
      id: d.id || '',
      name: d.place_name || '',
      address: d.address_name || '',
      road_address: d.road_address_name || '',
      lat: parseFloat(d.y || '0'),
      lng: parseFloat(d.x || '0'),
      phone: d.phone || '',
      distance: parseInt(d.distance || '0', 10) || 0,
      category: d.category_name || '',
    }))
  } catch {
    return []
  }
}

export async function searchNearbyBanks(lat: number, lng: number, radius = 3000): Promise<KakaoPlace[]> {
  const results = await searchNearbyPlaces({
    query: '은행', lat, lng, radius, category_group_code: 'BK9', size: 15,
  })
  const exclude = ['ATM', '365', '자동화', '무인']
  const filtered = results.filter(b => !exclude.some(kw => b.name.includes(kw)))
  return filtered.length > 0 ? filtered : results.slice(0, 5)
}

export async function searchNearbyPostOffices(lat: number, lng: number, radius = 3000): Promise<KakaoPlace[]> {
  return searchNearbyPlaces({ query: '우체국', lat, lng, radius, size: 10 })
}

/**
 * 통합 주소 검색 — 사용자가 결과를 고를 수 있도록 여러 후보를 반환.
 * 1) 카카오 주소 검색 (address.json) — 도로명/지번 주소 정확 매칭
 * 2) 카카오 키워드 검색 (keyword.json) — 건물명/상호/지명
 * 두 결과를 합쳐서 중복 제거 후 반환.
 */
export interface AddressSearchResult {
  label: string            // 표시용 대표 텍스트 (도로명 주소 우선)
  road_address: string
  address: string          // 지번 주소
  place_name?: string      // 장소명 (키워드 검색 시)
  lat: number
  lng: number
  postal_code?: string
  source: 'address' | 'keyword'
}

export async function searchAddresses(query: string, maxResults: number = 10): Promise<AddressSearchResult[]> {
  const q = query.trim()
  if (!q) return []

  const results: AddressSearchResult[] = []
  const seen = new Set<string>()

  const pushUnique = (r: AddressSearchResult) => {
    if (!r.lat || !r.lng) return
    const key = `${r.lat.toFixed(5)}_${r.lng.toFixed(5)}_${r.label}`
    if (seen.has(key)) return
    seen.add(key)
    results.push(r)
  }

  // 1) 주소 검색 (도로명/지번)
  try {
    const res = await httpGet<any>({
      url: `${LOCAL_BASE}/search/address.json`,
      params: { query: q, size: String(Math.min(maxResults, 30)) },
      headers: authHeaders(),
    })
    if (res.status === 200) {
      const docs = res.data?.documents || []
      for (const d of docs) {
        const road = d.road_address || {}
        const addr = d.address || {}
        const label = road.address_name || d.address_name || addr.address_name || ''
        pushUnique({
          label,
          road_address: road.address_name || '',
          address: addr.address_name || d.address_name || '',
          lat: parseFloat(d.y || '0'),
          lng: parseFloat(d.x || '0'),
          postal_code: road.zone_no || '',
          source: 'address',
        })
      }
    }
  } catch {
    /* ignore */
  }

  // 2) 키워드 검색 (장소명/건물명)
  try {
    const res = await httpGet<any>({
      url: `${LOCAL_BASE}/search/keyword.json`,
      params: { query: q, size: String(Math.min(maxResults, 15)) },
      headers: authHeaders(),
    })
    if (res.status === 200) {
      const docs = res.data?.documents || []
      for (const d of docs) {
        const label = d.place_name || d.road_address_name || d.address_name || ''
        pushUnique({
          label,
          road_address: d.road_address_name || '',
          address: d.address_name || '',
          place_name: d.place_name || '',
          lat: parseFloat(d.y || '0'),
          lng: parseFloat(d.x || '0'),
          source: 'keyword',
        })
      }
    }
  } catch {
    /* ignore */
  }

  return results.slice(0, maxResults)
}


export async function geocodeAddress(address: string): Promise<{
  address: string
  road_address: string
  lat: number
  lng: number
} | null> {
  if (!address.trim()) return null
  try {
    const res = await httpGet<any>({
      url: `${LOCAL_BASE}/search/address.json`,
      params: { query: address },
      headers: authHeaders(),
    })
    if (res.status !== 200) return null
    const docs = res.data?.documents || []
    if (docs.length === 0) {
      // 주소 검색 실패 → 키워드 검색 폴백
      const r2 = await httpGet<any>({
        url: `${LOCAL_BASE}/search/keyword.json`,
        params: { query: address, size: '1' },
        headers: authHeaders(),
      })
      if (r2.status !== 200) return null
      const docs2 = r2.data?.documents || []
      if (docs2.length === 0) return null
      const d = docs2[0]
      return {
        address: d.address_name || address,
        road_address: d.road_address_name || '',
        lat: parseFloat(d.y || '0'),
        lng: parseFloat(d.x || '0'),
      }
    }
    const d = docs[0]
    const road = d.road_address || {}
    return {
      address: d.address_name || address,
      road_address: road.address_name || '',
      lat: parseFloat(d.y || '0'),
      lng: parseFloat(d.x || '0'),
    }
  } catch {
    return null
  }
}

export async function reverseGeocode(lat: number, lng: number): Promise<{
  address: string
  road_address: string
  region_1depth: string
  region_2depth: string
  region_3depth: string
} | null> {
  try {
    const res = await httpGet<any>({
      url: `${LOCAL_BASE}/geo/coord2address.json`,
      params: { x: String(lng), y: String(lat) },
      headers: authHeaders(),
    })
    if (res.status !== 200) return null
    const docs = res.data?.documents || []
    if (docs.length === 0) return null
    const d = docs[0]
    const addr = d.address || {}
    const road = d.road_address || {}
    return {
      address: addr.address_name || '',
      road_address: road.address_name || '',
      region_1depth: addr.region_1depth_name || '',
      region_2depth: addr.region_2depth_name || '',
      region_3depth: addr.region_3depth_name || '',
    }
  } catch {
    return null
  }
}

export interface DirectionsResult {
  distance: number
  duration: number
  route_coords: { lat: number; lng: number }[]
}

export async function getDirections(
  originLng: number, originLat: number,
  destLng: number, destLat: number,
  waypoints?: [number, number][],
): Promise<DirectionsResult | null> {
  try {
    const params: Record<string, string> = {
      origin: `${originLng},${originLat}`,
      destination: `${destLng},${destLat}`,
      priority: 'RECOMMEND',
    }
    if (waypoints && waypoints.length > 0) {
      params.waypoints = waypoints.map(([lng, lat]) => `${lng},${lat}`).join('|')
    }
    const res = await httpGet<any>({
      url: NAVI_BASE,
      params,
      headers: authHeaders(),
    })
    if (res.status !== 200) return null
    const routes = res.data?.routes || []
    if (routes.length === 0) return null
    const route = routes[0]
    const summary = route.summary || {}
    const coords: { lat: number; lng: number }[] = []
    for (const section of route.sections || []) {
      for (const road of section.roads || []) {
        const vertexes = road.vertexes || []
        for (let i = 0; i < vertexes.length; i += 2) {
          if (i + 1 < vertexes.length) {
            coords.push({ lng: vertexes[i], lat: vertexes[i + 1] })
          }
        }
      }
    }
    return {
      distance: summary.distance || 0,
      duration: summary.duration || 0,
      route_coords: coords,
    }
  } catch {
    return null
  }
}

export async function getMultiStopRoute(
  start: [number, number],
  stops: [number, number][],
): Promise<DirectionsResult[]> {
  const segments: DirectionsResult[] = []
  let current = start
  for (const stop of stops) {
    const result = await getDirections(current[0], current[1], stop[0], stop[1])
    if (result) {
      segments.push(result)
    } else {
      segments.push({
        distance: 0,
        duration: 0,
        route_coords: [
          { lng: current[0], lat: current[1] },
          { lng: stop[0], lat: stop[1] },
        ],
      })
    }
    current = stop
  }
  return segments
}
